File: src/0142-Linked-List-Cycle-II/0142.cpp
#include <iostream>
#include <string>
#include <vector>
using namespace std;


//Definition for singly-linked list.
struct ListNode 
{
    int val;
    ListNode *next;
    ListNode(int x) : val(x), next(NULL) {}
};

static int x = []() {std::ios::sync_with_stdio(false); cin.tie(0); return 0; }();
class Solution 
{
public:
    ListNode *detectCycle(ListNode *head) 
    {
        if (head == nullptr) return nullptr;
        auto fast = head;
        auto slow = head;
        auto entry = head;
        while (fast->next and fast->next->next)
        {
            slow = slow->next;
            fast = fast->next->next;
            if (slow == fast)
            {
                while (slow != entry)
                {
                    slow = slow->next;
                    entry = entry->next;
                }
                return entry;
            }
        }
        return nullptr;
    }
};

File: src/0142-Linked-List-Cycle-II/0142.py
# Definition for singly-linked list.
class ListNode(object):
    def __init__(self, x):
        self.val = x
        self.next = None

class Solution:
    def findDuplicate(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        if len(nums) > 1:
            slow = nums[0]
            fast = nums[nums[0]]
            while slow != fast:
                slow = nums[slow]
                fast = nums[nums[fast]]
                
            entry = 0
            while entry != slow:
                entry = nums[entry]
                slow = nums[slow]
                
            return entry
        
        return -1